File: cpp/problem014.cpp
/*
    Copyright (c) 2015 - buyuk.dev@gmail.com
    ----------------------------------------
    Compute starting number uunder N which produces
    the longest Collatz sequence.
*/


#include <iostream>

using namespace std;

long collatz(long n) {

    long counter = 1;
    while (n>1) {
        if (n&1) {
            n = 3*n+1;
        } else {
            n = n >> 1;
        }
        counter ++;
    }
    return counter;
}


int main(int argc, char* argv[]) 
{
    ios_base::sync_with_stdio(0);

    long N;
    cin >> N;

    long max = 0;
    long num = -1;
    for(long i=1; i<=N; ++i) {
        long csl = collatz(i);
        if (max < csl) {
            max = csl;
            num = i;
        } 
    }

    cout << "collatz sequence for " 
         << num << " has length " 
         << max << endl;
    
    return 0;
}

File: python/problem014.py
#
# Copyright (c) 2016 - buyuk.dev@gmail.com
# ----------------------------------------
# Longest collatz sequence
#

def load(filename):

    data = 0
    with open(filename, "r") as dataFile:
        line = dataFile.readline()
        data = int(line)

    return data


def computeLength(n):

    counter = 1
    while n > 1:
        if n % 2 == 1:
            n = 3 * n + 1
        else:
            n = n // 2
        counter += 1

    return counter


def collatz(COL, n):

    if COL[n] is None:
        COL[n] = computeLength(n)

    return COL[n]


if __name__ == '__main__':

    n = load("data/problem014.dat")
    
    COL = [None] * (n + 1) 
    
    longest, best = (-1, -1)
    for i in range(n, 0, -1):
        length = collatz(COL, i)
        if longest < length:
            longest = length
            best = i

    print "collatz( ", best, " ) has length ", longest
